lib/src/rules/prefer_relative_imports.dart

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2019, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/file_system/file_system.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/ast.dart';
+import 'package:linter/src/rules/implementation_imports.dart';
+import 'package:yaml/yaml.dart';
+
+const _desc = r'Prefer relative imports for files in `lib/`.';
+
+const _details = r'''Prefer relative imports for files in `lib/`.
+
+When mixing relative and absolute imports it's possible to create confusion
+where the same member gets imported in two different ways. One way to avoid
+that is to ensure you consistently use relative imports for files withing the
+`lib/` directory.
+
+**GOOD:**
+
+```
+import 'bar.dart';
+...
+```
+
+**BAD:**
+
+```
+import 'package:my_package/bar.dart';
+...
+```
+
+''';
+
+class PreferRelativeImports extends LintRule implements NodeLintRule {
+  PreferRelativeImports()
+      : super(
+            name: 'prefer_relative_imports',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    final visitor = _Visitor(this);
+    registry.addCompilationUnit(this, visitor);
+    registry.addImportDirective(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final PreferRelativeImports rule;
+
+  bool isInLibFolder;
+  File pubspecFile;
+  Map parsedPubspec;
```

Comment:
In the future I guess we might consider storing this in `ast.dart` or somewhere so that other lints can take advantage of it?  Or: do you envision this something we will ultimately get from the analysis context? 

@bwilkerson: thoughts on that?

Reply:
I agree that this kind of support should be shared. I think it should be available from the analyzer (though I haven't yet thought about a concrete API for it) because we're likely to want it for hints as well.

Reply:
@bwilkerson: if you're keeping a running "technical debt" list, could you consider adding it there?   (Or suggesting how I can best capture the idea?)

Thanks!

Reply:
I don't really have such a list. I suppose an issue is probably best.

Reply:
👍 https://github.com/dart-lang/sdk/issues/38616